### PR TITLE
feat(connector): implement Reverse (VoidPostCapture) for PlaceToPay

### PR DESCRIPTION
**Pull request by 10X-GRACE**
## Summary

- Implements the Reverse (VoidPostCapture) flow for the PlaceToPay connector
- Uses PlaceToPay's POST /rest/gateway/transaction endpoint with action: "reverse" to perform post-capture payment reversals
- Passes the authorization code (stored in connector_feature_data from the original Authorize response) back to the PlaceToPay API, which is required for a successful reversal
- Also fixes PaymentFlowData::foreign_try_from for PaymentServiceReverseRequest to correctly extract connector_feature_data from the gRPC request (previously hardcoded to None)

## Changes

- placetopay.rs: Added VoidPC to create_all_prerequisites! and implemented macro_connector_implementation! block with get_headers and get_url
- transformers.rs: Added authorization: Option<Secret<String>> field to PlacetopayNextActionRequest (skipped when None); added TryFrom impl for VoidPC flow that extracts the authorization code from resource_common_data.connector_feature_data
- crates/types-traits/domain_types/src/types.rs: Fixed ForeignTryFrom for PaymentServiceReverseRequest to extract connector_feature_data from the gRPC request instead of hardcoding None

## Test plan

- [x] cargo build --package connector-integration passes
- [x] cargo build --bin grpc-server passes
- [x] gRPC types.PaymentService/Authorize returns CHARGED with connectorFeatureData
- [x] gRPC types.PaymentService/Reverse returns CHARGED with statusCode: 200 confirming post-capture reversal succeeded
- [x] Credential scrub: no API keys or secrets in diff

## Live gRPC Test Results

```
--- grpcurl: Authorize (Card, MANUAL capture) ---
grpcurl -plaintext \
  -import-path proto/ \
  -proto services.proto \
  -H 'x-connector: placetopay' \
  -H 'x-auth: body-key' \
  -H 'x-api-key: <REDACTED>' \
  -H 'x-key1: <REDACTED>' \
  -H 'x-merchant-id: DefaultMerchantId' \
  -H 'x-tenant-id: default' \
  -H 'x-request-id: auth-ptp-final-1778577538' \
  -d '{
    "merchant_transaction_id": "ptp_final_test_1778577538",
    "amount": {"minor_amount": 1000, "currency": "USD"},
    "payment_method": {"card": {"card_number": {"value": "4111111111111111"}, "card_exp_month": {"value": "03"}, "card_exp_year": {"value": "2030"}, "card_cvc": {"value": "737"}, "card_holder_name": {"value": "John Doe"}}},
    "capture_method": "MANUAL",
    "address": {"billing_address": {"first_name": {"value": "John"}, "last_name": {"value": "Doe"}, "line1": {"value": "123 Test St"}, "city": {"value": "Bogota"}, "zip_code": {"value": "110111"}, "country_alpha2_code": "CO"}},
    "auth_type": "NO_THREE_DS",
    "browser_info": {"user_agent": "Mozilla/5.0 (X11; Linux x86_64) Chrome/120.0", "accept_header": "text/html,application/xhtml+xml", "language": "en-US", "color_depth": 24, "screen_height": 1080, "screen_width": 1920, "time_zone_offset_minutes": -300, "java_enabled": false, "java_script_enabled": true, "ip_address": "192.168.1.1"},
    "description": "Test for PlaceToPay Reverse",
    "return_url": "https://example.com/return"
  }' \
  localhost:8903 types.PaymentService/Authorize

Response:
{
  "connectorTransactionId": "1599767984",
  "status": "CHARGED",
  "statusCode": 200,
  "connectorFeatureData": {
    "value": "\"000000\""
  }
}

--- grpcurl: Reverse (VoidPostCapture) ---
grpcurl -plaintext \
  -import-path proto/ \
  -proto services.proto \
  -H 'x-connector: placetopay' \
  -H 'x-auth: body-key' \
  -H 'x-api-key: <REDACTED>' \
  -H 'x-key1: <REDACTED>' \
  -H 'x-merchant-id: DefaultMerchantId' \
  -H 'x-tenant-id: default' \
  -H 'x-request-id: rev-ptp-final-1778577558' \
  -d '{
    "merchant_reverse_id": "ptp_reverse_final_001",
    "connector_transaction_id": "1599767984",
    "connector_feature_data": {"value": "\"000000\""}
  }' \
  localhost:8903 types.PaymentService/Reverse

Response:
{
  "connectorTransactionId": "1599767985",
  "status": "CHARGED",
  "statusCode": 200
}

PlaceToPay endpoint: POST https://test.placetopay.com/rest/gateway/transaction
Actual request body sent: {"auth":{...},"internalReference":1599767984,"action":"reverse","authorization":"000000"}
Reversal accepted by gateway (HTTP 200, APPROVED from PlaceToPay sandbox)
```

Status CHARGED on Reverse maps to PaymentStatus::Charged — PlaceToPay returns APPROVED which maps to AttemptStatus::Charged. Reversal was accepted by the gateway with HTTP 200.

Generated with Claude Code

### CODE REVIEW

**Comment by 10X-GRACE** [comment]
## Updated gRPC Test Results — PostCaptureVoidResponse

Branch rebased on remote tip and `PostCaptureVoidResponse` wired in for the VoidPC (Reverse) flow.

Key changes in this update:
- Cherry-picked two framework commits from `feat/grace-reverse-flow`:
  - `feat(framework): add PostCaptureVoidResponse to PaymentsResponseData`
  - `fix(framework): use PaymentStatus::Pending not Processing`
- Added dedicated `PlacetopayVoidPcResponse` newtype (transparent wrapper) with its own `TryFrom` returning `PostCaptureVoidResponse`
- Fixed rebase conflict: removed duplicate `connector_feature_data` extraction that caused `E0790`

<details>
<summary>Authorize → Reverse (credentials redacted)</summary>

```
=== AUTHORIZE ===
{
  "connectorTransactionId": "[REDACTED]",
  "status": "CHARGED",
  "statusCode": 200,
  "connectorFeatureData": { "value": "[REDACTED]" }
}

=== REVERSE ===
{
  "connectorTransactionId": "[REDACTED]",
  "status": "VOIDED_POST_CAPTURE",
  "statusCode": 200,
  "merchantReverseId": "[REDACTED]"
}
```

</details>

- Authorize: `CHARGED` (status_code 200) ✓
- Reverse: `VOIDED_POST_CAPTURE` (status_code 200) ✓ — `PostCaptureVoidResponse` with `PostCaptureVoidStatus::Succeeded`

**Comment by hs-agent** [comment]
@prism-review-bot please review this pr

File: crates/integrations/connector-integration/src/connectors/placetopay/transformers.rs [hunk 1]
```diff
@@ -250,6 +250,61 @@ impl<T: PaymentMethodDataTypes + Debug + Sync + Send + 'static + Serialize>
...
+            >,
+            T,
+        >,
+    ) -> Result<Self, Self::Error> {
+        let auth = PlacetopayAuth::try_from(&item.router_data.connector_config)?;
+        let internal_reference = item
+            .router_data
+            .request
+            .connector_transaction_id
+            .parse::<u64>()
```
Comment (hs-agent) [inline crates/integrations/connector-integration/src/connectors/placetopay/transformers.rs:288]: :warning: The connector_transaction_id parsing assumes numeric-only IDs via parse::<u64>(). If Placetopay returns alphanumeric transaction IDs, this will fail with RequestEncodingFailed.

Verify with Placetopay documentation whether their internalReference is guaranteed to be numeric.

File: crates/integrations/connector-integration/src/connectors/placetopay/transformers.rs [hunk 2]
```diff
@@ -250,6 +250,61 @@ impl<T: PaymentMethodDataTypes + Debug + Sync + Send + 'static + Serialize>
...
+                PaymentsCancelPostCaptureData,
+                PaymentsResponseData,
+            >,
+            T,
+        >,
+    > for PlacetopayNextActionRequest
+{
+    type Error = error_stack::Report<IntegrationError>;
+    fn try_from(
+        item: PlacetopayRouterData<
```
Comment (hs-agent) [inline crates/integrations/connector-integration/src/connectors/placetopay/transformers.rs:273]: :bulb: The authorization field is conditionally populated from connector_feature_data for VoidPC flows, but set to None for all other flows.

Confirm this is intentional: does Placetopay accept requests without an authorization code for Capture/Void actions?

File: crates/integrations/connector-integration/src/connectors/placetopay/transformers.rs [hunk 3]
```diff
@@ -331,6 +386,59 @@ impl<F, T> TryFrom<ResponseRouterData<PlacetopayPaymentsResponse, Self>>
...
+            PlacetopayTransactionStatus::Pending
+            | PlacetopayTransactionStatus::PendingValidation
+            | PlacetopayTransactionStatus::PendingProcess => {
+                common_enums::PostCaptureVoidStatus::Pending
+            }
+        };
+        let attempt_status = match post_capture_void_status {
+            common_enums::PostCaptureVoidStatus::Succeeded => {
+                common_enums::AttemptStatus::VoidedPostCapture
+            }
```
Comment (hs-agent) [inline crates/integrations/connector-integration/src/connectors/placetopay/transformers.rs]: :mag: Consider adding a test case for the VoidPC flow status mapping edge cases.

The field probe shows the flow is supported, but a unit test validating each PlacetopayTransactionStatus mapping would prevent regressions.

**Review by hs-agent — DISMISSED** [review]
**Verdict:** :arrows_counterclockwise: Request Changes

:warning: 1 warning · :bulb: 1 suggestion · :mag: 1 nitpick

**Classification:** connector-flow-addition · secondary: none · metadata: none
**Risk:** medium

This PR adds the VoidPC (Reverse) flow to the Placetopay connector. The implementation follows the established patterns for new flows: macro-based request/response construction, proper status mapping, and appropriate field probe updates.

**Key concerns:**
1. **Numeric transaction ID assumption**: The `parse::<u64>()` for `connector_transaction_id` assumes Placetopay only returns numeric internal references. This needs verification.
2. **Authorization field handling**: The conditional population of `authorization` for VoidPC but not other flows raises questions about API requirements for Capture/Void actions.
3. **Test coverage**: Consider adding unit tests for the status mapping logic to prevent regressions.

File: crates/integrations/connector-integration/src/connectors/placetopay/transformers.rs [hunk 4]
```diff
@@ -331,6 +386,46 @@ impl<F, T> TryFrom<ResponseRouterData<PlacetopayPaymentsResponse, Self>>
...
+            | PlacetopayTransactionStatus::PendingValidation
+            | PlacetopayTransactionStatus::PendingProcess => {
+                common_enums::PostCaptureVoidStatus::Pending
+            }
+        };
+        Ok(Self {
+            response: Ok(PaymentsResponseData::PostCaptureVoidResponse {
+                post_capture_void_status,
+                connector_reference_id: Some(inner.internal_reference.to_string()),
+                description: None,
```
Comment (AkshayaFoiger) [inline crates/integrations/connector-integration/src/connectors/placetopay/transformers.rs]: Incase of failed response, populate the description

Comment (10X-GRACE) [reply crates/integrations/connector-integration/src/connectors/placetopay/transformers.rs]: Addressed in 1bc277792 — `PlacetopayStatusResponse` now also deserializes the optional `message` and `reason` fields, and `description` is populated from `status.message` (falling back to `status.reason`) when `post_capture_void_status.is_post_capture_void_failure()` is true. Mirrors the worldpayvantiv pattern at worldpayvantiv/transformers.rs:2513-2515.

File: crates/integrations/connector-integration/src/connectors/placetopay/transformers.rs (hunk 1 above)
Comment (10X-GRACE) [reply crates/integrations/connector-integration/src/connectors/placetopay/transformers.rs:288]: This matches PlaceToPay's API contract: `internalReference` is documented as a numeric identifier and is typed as `u64` throughout this connector (Capture transformers.rs:240, Void transformers.rs:284, PSync, Refund, RSync — all via `parse::<u64>()`) and across the upstream hyperswitch reference (`hyperswitch_connectors/.../placetopay/transformers.rs` lines 261, 302, 386, 470, 494). Using the same pattern for VoidPC keeps the connector internally consistent. Live test against the PlaceToPay sandbox returned `1599773055` / `1599773056` — both pure numeric.

File: crates/integrations/connector-integration/src/connectors/placetopay/transformers.rs (hunk 2 above)
Comment (10X-GRACE) [reply crates/integrations/connector-integration/src/connectors/placetopay/transformers.rs:273]: Intentional and verified. In PlaceToPay's `/rest/gateway/transaction` API, the `authorization` field is only required when `action == reverse` (refunds and post-capture void). For `Checkout` (Capture) and `Void` actions the field must be absent. Hyperswitch reference confirms this: `PlacetopayRefundRequest` (action=Reverse) has `authorization: Option<String>` while `PlacetopayNextActionRequest` used by Capture/Void omits the field entirely (`hyperswitch_connectors/.../placetopay/transformers.rs:490-543`). Our struct uses a single `Option<Secret<String>>` field with `#[serde(skip_serializing_if = "Option::is_none")]`, so when `None` the key is omitted from the wire payload — wire-equivalent to hyperswitch's approach.

**Review by AkshayaFoiger — APPROVED** [review]

**Review by AmitsinghTanwar007 — APPROVED** [review]
